# ETL Pipeline Preparation
Follow the instructions below to help you create your ETL pipeline.
### 1. Import libraries and load datasets.
- Import Python libraries
- Load `messages.csv` into a dataframe and inspect the first few lines.
- Load `categories.csv` into a dataframe and inspect the first few lines.

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# load messages dataset
messages = pd.read_csv("messages.csv")
messages.head()

id                                            message  \
0   2  Weather update - a cold front from Cuba that c...   
1   7            Is the Hurricane over or is it not over   
2   8                    Looking for someone but no name   
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...   
4  12  says: west side of Haiti, rest of the country ...   

                                            original   genre  
0  Un front froid se retrouve sur Cuba ce matin. ...  direct  
1                 Cyclone nan fini osinon li pa fini  direct  
2  Patnm, di Maryani relem pou li banm nouvel li ...  direct  
3  UN reports Leogane 80-90 destroyed. Only Hospi...  direct  
4  facade ouest d Haiti et le reste du pays aujou...  direct

In [3]:
# load categories dataset
categories = pd.read_csv("categories.csv")
categories.head()

id                                         categories
0   2  related-1;request-0;offer-0;aid_related-0;medi...
1   7  related-1;request-0;offer-0;aid_related-1;medi...
2   8  related-1;request-0;offer-0;aid_related-0;medi...
3   9  related-1;request-1;offer-0;aid_related-1;medi...
4  12  related-1;request-0;offer-0;aid_related-0;medi...

### 2. Merge datasets.
- Merge the messages and categories datasets using the common id
- Assign this combined dataset to `df`, which will be cleaned in the following steps

In [4]:
# merge datasets
df = pd.merge(messages, categories, on="id", how="left")
df.head()

id                                            message  \
0   2  Weather update - a cold front from Cuba that c...   
1   7            Is the Hurricane over or is it not over   
2   8                    Looking for someone but no name   
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...   
4  12  says: west side of Haiti, rest of the country ...   

                                            original   genre  \
0  Un front froid se retrouve sur Cuba ce matin. ...  direct   
1                 Cyclone nan fini osinon li pa fini  direct   
2  Patnm, di Maryani relem pou li banm nouvel li ...  direct   
3  UN reports Leogane 80-90 destroyed. Only Hospi...  direct   
4  facade ouest d Haiti et le reste du pays aujou...  direct   

                                          categories  
0  related-1;request-0;offer-0;aid_related-0;medi...  
1  related-1;request-0;offer-0;aid_related-1;medi...  
2  related-1;request-0;offer-0;aid_related-0;medi...  
3  related-1;request-1;offer-0;aid_related-1;medi...  
4  related-1;request-0;offer-0;aid_related-0;medi...

### 3. Split `categories` into separate category columns.
- Split the values in the `categories` column on the `;` character so that each value becomes a separate column. You'll find [this method](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.Series.str.split.html) very helpful! Make sure to set `expand=True`.
- Use the first row of categories dataframe to create column names for the categories data.
- Rename columns of `categories` with new column names.

In [5]:
# create a dataframe of the 36 individual category columns
categories = df["categories"].str.split(";", expand=True)
categories.head()

0          1        2              3               4   \
0  related-1  request-0  offer-0  aid_related-0  medical_help-0   
1  related-1  request-0  offer-0  aid_related-1  medical_help-0   
2  related-1  request-0  offer-0  aid_related-0  medical_help-0   
3  related-1  request-1  offer-0  aid_related-1  medical_help-0   
4  related-1  request-0  offer-0  aid_related-0  medical_help-0   

                   5                    6           7           8   \
0  medical_products-0  search_and_rescue-0  security-0  military-0   
1  medical_products-0  search_and_rescue-0  security-0  military-0   
2  medical_products-0  search_and_rescue-0  security-0  military-0   
3  medical_products-1  search_and_rescue-0  security-0  military-0   
4  medical_products-0  search_and_rescue-0  security-0  military-0   

              9   ...             26                      27  \
0  child_alone-0  ...  aid_centers-0  other_infrastructure-0   
1  child_alone-0  ...  aid_centers-0  other_infrastructure-0   
2  child_alone-0  ...  aid_centers-0  other_infrastructure-0   
3  child_alone-0  ...  aid_centers-0  other_infrastructure-0   
4  child_alone-0  ...  aid_centers-0  other_infrastructure-0   

                  28        29       30      31            32      33  \
0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
1  weather_related-1  floods-0  storm-1  fire-0  earthquake-0  cold-0   
2  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
3  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
4  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   

                34               35  
0  other_weather-0  direct_report-0  
1  other_weather-0  direct_report-0  
2  other_weather-0  direct_report-0  
3  other_weather-0  direct_report-0  
4  other_weather-0  direct_report-0  

[5 rows x 36 columns]

In [6]:
# select the first row of the categories dataframe

row = categories.loc[0, :]

# use this row to extract a list of new column names for categories.
# one way is to apply a lambda function that takes everything
# up to the second to last character of each string with slicing
category_colnames = row.apply(lambda x: x[:-2])
print(category_colnames)

0                    related
1                    request
2                      offer
3                aid_related
4               medical_help
5           medical_products
6          search_and_rescue
7                   security
8                   military
9                child_alone
10                     water
11                      food
12                   shelter
13                  clothing
14                     money
15            missing_people
16                  refugees
17                     death
18                 other_aid
19    infrastructure_related
20                 transport
21                 buildings
22               electricity
23                     tools
24                 hospitals
25                     shops
26               aid_centers
27      other_infrastructure
28           weather_related
29                    floods
30                     storm
31                      fire
32                earthquake
33                      cold
34            

In [7]:
# rename the columns of `categories`
categories.columns = category_colnames
categories.head()

0    related    request    offer    aid_related    medical_help  \
0  related-1  request-0  offer-0  aid_related-0  medical_help-0   
1  related-1  request-0  offer-0  aid_related-1  medical_help-0   
2  related-1  request-0  offer-0  aid_related-0  medical_help-0   
3  related-1  request-1  offer-0  aid_related-1  medical_help-0   
4  related-1  request-0  offer-0  aid_related-0  medical_help-0   

0    medical_products    search_and_rescue    security    military  \
0  medical_products-0  search_and_rescue-0  security-0  military-0   
1  medical_products-0  search_and_rescue-0  security-0  military-0   
2  medical_products-0  search_and_rescue-0  security-0  military-0   
3  medical_products-1  search_and_rescue-0  security-0  military-0   
4  medical_products-0  search_and_rescue-0  security-0  military-0   

0    child_alone  ...    aid_centers    other_infrastructure  \
0  child_alone-0  ...  aid_centers-0  other_infrastructure-0   
1  child_alone-0  ...  aid_centers-0  other_infrastructure-0   
2  child_alone-0  ...  aid_centers-0  other_infrastructure-0   
3  child_alone-0  ...  aid_centers-0  other_infrastructure-0   
4  child_alone-0  ...  aid_centers-0  other_infrastructure-0   

0    weather_related    floods    storm    fire    earthquake    cold  \
0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
1  weather_related-1  floods-0  storm-1  fire-0  earthquake-0  cold-0   
2  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
3  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
4  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   

0    other_weather    direct_report  
0  other_weather-0  direct_report-0  
1  other_weather-0  direct_report-0  
2  other_weather-0  direct_report-0  
3  other_weather-0  direct_report-0  
4  other_weather-0  direct_report-0  

[5 rows x 36 columns]

### 4. Convert category values to just numbers 0 or 1.
- Iterate through the category columns in df to keep only the last character of each string (the 1 or 0). For example, `related-0` becomes `0`, `related-1` becomes `1`. Convert the string to a numeric value.
- You can perform [normal string actions on Pandas Series](https://pandas.pydata.org/pandas-docs/stable/text.html#indexing-with-str), like indexing, by including `.str` after the Series. You may need to first convert the Series to be of type string, which you can do with `astype(str)`.

In [8]:
for column in categories:
    # set each value to be the last character of the string
    categories[column] = categories[column].str[-1].astype(int)
    # convert column from string to numeric
    #categories[column] = categories[column].astype(int)
categories.head()

0  related  request  offer  aid_related  medical_help  medical_products  \
0        1        0      0            0             0                 0   
1        1        0      0            1             0                 0   
2        1        0      0            0             0                 0   
3        1        1      0            1             0                 1   
4        1        0      0            0             0                 0   

0  search_and_rescue  security  military  child_alone  ...  aid_centers  \
0                  0         0         0            0  ...            0   
1                  0         0         0            0  ...            0   
2                  0         0         0            0  ...            0   
3                  0         0         0            0  ...            0   
4                  0         0         0            0  ...            0   

0  other_infrastructure  weather_related  floods  storm  fire  earthquake  \
0                     0                0       0      0     0           0   
1                     0                1       0      1     0           0   
2                     0                0       0      0     0           0   
3                     0                0       0      0     0           0   
4                     0                0       0      0     0           0   

0  cold  other_weather  direct_report  
0     0              0              0  
1     0              0              0  
2     0              0              0  
3     0              0              0  
4     0              0              0  

[5 rows x 36 columns]

* Verify all transformed data are binary (0 and 1)

In [9]:
# show value counts of each category column
df_value_counts = categories.apply(lambda x: x.value_counts()).T
df_value_counts[df_value_counts[2].notna()]

0        1      2
0                              
related  6140.0  20042.0  204.0

* There are 204 entires/messages in the original dataset marked as 2 in the "related" category. Let's take a closer look at those messsages

In [10]:
# show message where `related` column has a value of 2
df[categories["related"] == 2]

id                                            message  \
117      146  Dans la zone de Saint Etienne la route de Jacm...   
221      263  . .. i with limited means. Certain patients co...   
307      373  The internet caf Net@le that's by the Dal road...   
462      565  Bonsoir, on est a bon repos aprs la compagnie ...   
578      700  URGENT CRECHE ORPHANAGE KAY TOUT TIMOUN CROIX ...   
...      ...                                                ...   
20465  23662       transferred to a sanitary landfill site by a   
20636  23861  Families also have solar lamps which can be re...   
22481  25874  Read the [full blog post](http://www.odi.org.u...   
23537  27071  Actualmente e independientemente de la ayuda d...   
25385  29126  Mali's former President Amadou Toumani Touré -...   

                                                original   genre  \
117    Nan zon st. etine rout jakmel la bloke se mize...  direct   
221    t avec des moyens limites. Certains patients v...  direct   
307    Cyber cafe net@le ki chita rout de dal tou pr ...  direct   
462    Bonswa nou nan bon repo apri teleko nan wout t...  direct   
578    r et Salon Furterer. mwen se yon Cosmtologue. ...  direct   
...                                                  ...     ...   
20465                                                NaN    news   
20636                                                NaN    news   
22481                                                NaN    news   
23537                                                NaN    news   
25385                                                NaN    news   

                                              categories  
117    related-2;request-0;offer-0;aid_related-0;medi...  
221    related-2;request-0;offer-0;aid_related-0;medi...  
307    related-2;request-0;offer-0;aid_related-0;medi...  
462    related-2;request-0;offer-0;aid_related-0;medi...  
578    related-2;request-0;offer-0;aid_related-0;medi...  
...                                                  ...  
20465  related-2;request-0;offer-0;aid_related-0;medi...  
20636  related-2;request-0;offer-0;aid_related-0;medi...  
22481  related-2;request-0;offer-0;aid_related-0;medi...  
23537  related-2;request-0;offer-0;aid_related-0;medi...  
25385  related-2;request-0;offer-0;aid_related-0;medi...  

[204 rows x 5 columns]

* Messages flagged as "related-2" seem to be mostly in non-English languages or references to external sources (i.e. news). Since these messages only account for a very small portion of the raw data and are unreliable, we will exclude them from our analysis.

In [11]:
# generate row index for rows where`related` column marked as 2
index_drop = categories[categories["related"] == 2].index
# drop the marked rows accordingly in both df and categories dataframes
df.drop(index_drop, inplace=True)
categories.drop(index_drop, inplace=True)

- The underlying data (0s and 1s) for each category is very imbalanced as shown in the table below
- 100% of the data in the `child_alone` category is 0. This will not add any meaningful information to train the machine learning model later on. Thus, we should drop the `child_alone` column before doing machine learning

In [12]:
# show value counts for each category
value_counts = categories.apply(lambda x: x.value_counts()).T
value_counts["% of value = 1"] = value_counts[1]/len(categories)
value_counts

0        1  % of value = 1
0                                                       
related                  6140.0  20042.0        0.765488
request                 21669.0   4513.0        0.172370
offer                   26061.0    121.0        0.004621
aid_related             15228.0  10954.0        0.418379
medical_help            24083.0   2099.0        0.080170
medical_products        24863.0   1319.0        0.050378
search_and_rescue       25457.0    725.0        0.027691
security                25711.0    471.0        0.017989
military                25319.0    863.0        0.032962
child_alone             26182.0      NaN             NaN
water                   24498.0   1684.0        0.064319
food                    23226.0   2956.0        0.112902
shelter                 23840.0   2342.0        0.089451
clothing                25772.0    410.0        0.015660
money                   25576.0    606.0        0.023146
missing_people          25881.0    301.0        0.011496
refugees                25302.0    880.0        0.033611
death                   24978.0   1204.0        0.045986
other_aid               22718.0   3464.0        0.132305
infrastructure_related  24473.0   1709.0        0.065274
transport               24970.0   1212.0        0.046291
buildings               24837.0   1345.0        0.051371
electricity             25644.0    538.0        0.020548
tools                   26023.0    159.0        0.006073
hospitals               25899.0    283.0        0.010809
shops                   26062.0    120.0        0.004583
aid_centers             25873.0    309.0        0.011802
other_infrastructure    25027.0   1155.0        0.044114
weather_related         18839.0   7343.0        0.280460
floods                  24005.0   2177.0        0.083149
storm                   23718.0   2464.0        0.094110
fire                    25900.0    282.0        0.010771
earthquake              23721.0   2461.0        0.093996
cold                    25649.0    533.0        0.020357
other_weather           24803.0   1379.0        0.052670
direct_report           21069.0   5113.0        0.195287

### 5. Replace `categories` column in `df` with new category columns.
- Drop the categories column from the df dataframe since it is no longer needed.
- Concatenate df and categories data frames.

In [13]:
# drop the original categories column from `df`
df.drop(labels="categories", axis=1, inplace=True)

df.head()

id                                            message  \
0   2  Weather update - a cold front from Cuba that c...   
1   7            Is the Hurricane over or is it not over   
2   8                    Looking for someone but no name   
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...   
4  12  says: west side of Haiti, rest of the country ...   

                                            original   genre  
0  Un front froid se retrouve sur Cuba ce matin. ...  direct  
1                 Cyclone nan fini osinon li pa fini  direct  
2  Patnm, di Maryani relem pou li banm nouvel li ...  direct  
3  UN reports Leogane 80-90 destroyed. Only Hospi...  direct  
4  facade ouest d Haiti et le reste du pays aujou...  direct

In [14]:
# concatenate the original dataframe with the new `categories` dataframe
df = pd.concat([df, categories], axis=1)
df.head()

id                                            message  \
0   2  Weather update - a cold front from Cuba that c...   
1   7            Is the Hurricane over or is it not over   
2   8                    Looking for someone but no name   
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...   
4  12  says: west side of Haiti, rest of the country ...   

                                            original   genre  related  \
0  Un front froid se retrouve sur Cuba ce matin. ...  direct        1   
1                 Cyclone nan fini osinon li pa fini  direct        1   
2  Patnm, di Maryani relem pou li banm nouvel li ...  direct        1   
3  UN reports Leogane 80-90 destroyed. Only Hospi...  direct        1   
4  facade ouest d Haiti et le reste du pays aujou...  direct        1   

   request  offer  aid_related  medical_help  medical_products  ...  \
0        0      0            0             0                 0  ...   
1        0      0            1             0                 0  ...   
2        0      0            0             0                 0  ...   
3        1      0            1             0                 1  ...   
4        0      0            0             0                 0  ...   

   aid_centers  other_infrastructure  weather_related  floods  storm  fire  \
0            0                     0                0       0      0     0   
1            0                     0                1       0      1     0   
2            0                     0                0       0      0     0   
3            0                     0                0       0      0     0   
4            0                     0                0       0      0     0   

   earthquake  cold  other_weather  direct_report  
0           0     0              0              0  
1           0     0              0              0  
2           0     0              0              0  
3           0     0              0              0  
4           0     0              0              0  

[5 rows x 40 columns]

### 6. Remove duplicates.
- Check how many duplicates are in this dataset.
- Drop the duplicates.
- Confirm duplicates were removed.

In [15]:
# check number of duplicates
df.duplicated().sum()

154

In [16]:
# show duplicated rows
df.loc[df.duplicated(keep=False), :].head()

id                                            message  \
162  202  ?? port au prince ?? and food. they need gover...   
163  202  ?? port au prince ?? and food. they need gover...   
164  202  ?? port au prince ?? and food. they need gover...   
165  202  ?? port au prince ?? and food. they need gover...   
713  862  What is the address of the radio station? I as...   

                                              original   genre  related  \
162  p bay pap la syen ak manje. Yo bezwen ed gouve...  direct        1   
163  p bay pap la syen ak manje. Yo bezwen ed gouve...  direct        1   
164  p bay pap la syen ak manje. Yo bezwen ed gouve...  direct        1   
165  p bay pap la syen ak manje. Yo bezwen ed gouve...  direct        1   
713  Ki adres radyo a? Paske m bezwen al depoze dos...  direct        0   

     request  offer  aid_related  medical_help  medical_products  ...  \
162        1      0            1             0                 0  ...   
163        1      0            1             0                 0  ...   
164        1      0            1             0                 0  ...   
165        1      0            1             0                 0  ...   
713        0      0            0             0                 0  ...   

     aid_centers  other_infrastructure  weather_related  floods  storm  fire  \
162            0                     0                0       0      0     0   
163            0                     0                0       0      0     0   
164            0                     0                0       0      0     0   
165            0                     0                0       0      0     0   
713            0                     0                0       0      0     0   

     earthquake  cold  other_weather  direct_report  
162           0     0              0              0  
163           0     0              0              0  
164           0     0              0              0  
165           0     0              0              0  
713           0     0              0              0  

[5 rows x 40 columns]

In [17]:
# drop duplicates
df.drop_duplicates(keep="first", inplace=True)

In [18]:
# check number of duplicates
num_dups = df.duplicated().sum()
print(f"The number of duplicates in the dataframe now is {num_dups}")

The number of duplicates in the dataframe now is 0


### 7. Save the clean dataset into an sqlite database.
You can do this with pandas [`to_sql` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.to_sql.html) combined with the SQLAlchemy library. Remember to import SQLAlchemy's `create_engine` in the first cell of this notebook to use it below.

In [19]:
# create database
engine = create_engine("sqlite:///disaster_response.db")
df.to_sql("messages", engine, index=False, if_exists="replace")

### 8. Use this notebook to complete `etl_pipeline.py`
Use the template file attached in the Resources folder to write a script that runs the steps above to create a database based on new datasets specified by the user. Alternatively, you can complete `etl_pipeline.py` in the classroom on the `Project Workspace IDE` coming later.